### PR TITLE
Map ++ without CBF should use the same optimised path as ++ with CBF

### PR DESCRIPTION
**Pull request by mkeskells**
also optimise for small map addition

### CODE REVIEW

**Review by diesalbla — APPROVED** [review]

File: src/library/scala/collection/immutable/Map.scala [hunk 1]
```diff
@@ -151,6 +150,20 @@ object Map extends ImmutableMapFactory[Map] {
       if (key == key1) new Map1(key1, value)
       else new Map2(key1, value1, key, value)
     def + [V1 >: V](kv: (K, V1)): Map[K, V1] = updated(kv._1, kv._2)
+    override def ++[V1 >: V](xs: GenTraversableOnce[(K, V1)]): Map[K, V1] = ++[(K, V1), Map[K, V1]](xs)(Map.canBuildFrom[K, V1])
+    override def ++[B >: (K, V), That](that: GenTraversableOnce[B])(implicit bf: CanBuildFrom[Map[K, V], B, That]): That = {
+      if (isMapCBF(bf)) that match {
+        case m if m.asInstanceOf[AnyRef] eq EmptyMap => this.asInstanceOf[That]
```
Comment (lrytz) [inline src/library/scala/collection/immutable/Map.scala]: `case EmptyMap`?

File: src/library/scala/collection/immutable/Map.scala [hunk 2]
```diff
@@ -204,6 +217,22 @@ object Map extends ImmutableMapFactory[Map] {
       else if (key == key2) new Map2(key1, value1, key2, value)
       else new Map3(key1, value1, key2, value2, key, value)
     def + [V1 >: V](kv: (K, V1)): Map[K, V1] = updated(kv._1, kv._2)
+    override def ++[V1 >: V](xs: GenTraversableOnce[(K, V1)]): Map[K, V1] = ++[(K, V1), Map[K, V1]](xs)(Map.canBuildFrom[K, V1])
+    override def ++[B >: (K, V), That](that: GenTraversableOnce[B])(implicit bf: CanBuildFrom[Map[K, V], B, That]): That = {
+      if (isMapCBF(bf)) that match {
+        case m if m.asInstanceOf[AnyRef] eq EmptyMap => this.asInstanceOf[That]
+        case m: Map1[K, V] => (if(m.contains(key1)) m else m.addTo(this)).asInstanceOf[That]
```
Comment (lrytz) [inline src/library/scala/collection/immutable/Map.scala]: Is this right? we're in `Map2`, why does it check only for `key1`?

Also this seems to violate a consistency rule
https://github.com/scala/scala/blob/aeee8f0b2f9f970d07dd33fdb49e93fa1140e103/test/junit/scala/collection/SetMapRulesTest.scala#L30

File: src/library/scala/collection/immutable/Map.scala (hunk 1 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/Map.scala]: that would check if `that == EmptyMap`, and may involve `size` `subSetOf` etc, it was just to be able to make a quick noop check

I wish there was a shorthand for the eq check in a case match though

File: src/library/scala/collection/immutable/Map.scala (hunk 2 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/Map.scala]: the key1 check is completely wrong, and should be in line with the other MapN classes
I used to have some code like that, but it all got too complex. Maybe it was a bad rebase I did 😲 removed now 

I think it is in line with the consistency rule

m.addTo(this)  calls Map1#addTo, which updates this Map2 by adding calling updated

so - it now reads
```
case m: Map1[K, V] => m.addTo(this).asInstanceOf[That]
case m: Map2[K, V] => m.addTo(this).asInstanceOf[That]
```
which is effectively 
```
case m: Map1[K, V] => (this.updated(m.key1, m.value1)).asInstanceOf[That]
case m: Map2[K, V] => (this.updated(m.key1, m.value1).updated(m.key2, m.value2)).asInstanceOf[That]
```
but with a little less code

File: src/library/scala/collection/immutable/Map.scala (hunk 1 above)
Comment (lrytz) [reply src/library/scala/collection/immutable/Map.scala]: OK. `case m: AnyRef if m eq EmptyMap` would be slightly cleaner. Also the existing `case EmptyMap` in https://github.com/scala/scala/blob/2838d0454541ed0681fa0d9b47becbba6e7df159/src/library/scala/collection/immutable/Map.scala#L120 could be made the same.

**Review by lrytz — COMMENTED** [review]
LGTM, after squashing the changes.

**Comment by mkeskells** [comment]
> LGTM, after squashing the changes.

squashed and added 
`case m: AnyRef if m eq EmptyMap` if the 5 places suggested

File: src/library/scala/collection/immutable/Map.scala (hunk 1 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/Map.scala]: will pop those in another PR

**Review by lrytz — APPROVED** [review]

**Pull request merged**